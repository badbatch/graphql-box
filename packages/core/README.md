# @graphql-box/core

The GraphQL Box core module.

[![npm version](https://badge.fury.io/js/%40graphql-box%2Fcore.svg)](https://badge.fury.io/js/%40graphql-box%2Fcore)
[![License: MIT](https://img.shields.io/badge/License-MIT-yellow.svg)](LICENSE)

## Installation

```bash
yarn add @graphql-box/core
```

## Usage

Read the GraphQL Box [usage guide](../../README.md#usage).

## Documentation

Read the module API [documentation](docs/README.md).
