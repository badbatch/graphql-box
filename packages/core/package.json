{
  "name": "@graphql-box/core",
  "version": "4.1.6",
  "description": "The GraphQL Box core module.",
  "author": "Dylan Aubrey <dylanaubrey@gmail.com>",
  "license": "MIT",
  "homepage": "https://github.com/badbatch/graphql-box",
  "bugs": {
    "url": "https://github.com/badbatch/graphql-box/issues"
  },
  "repository": {
    "type": "git",
    "url": "https://github.com/badbatch/graphql-box",
    "directory": "packages/core"
  },
  "publishConfig": {
    "access": "public"
  },
  "main": "lib/main/index.js",
  "module": "lib/module/index.mjs",
  "browser": "lib/browser/index.js",
  "types": "lib/types/index.d.ts",
  "keywords": [
    "graphql-box",
    "graphql",
    "client",
    "server",
    "isomorphic"
  ],
  "scripts": {
    "compile": "concurrently npm:compile:*",
    "compile-watch": "concurrently npm:compile-watch:*",
    "compile-watch:main": "yarn run compile:main --watch",
    "compile-watch:types": "yarn run compile:types --watch",
    "compile:browser": "cross-env BABEL_ENV=browser rollup -c ../../rollup.config.js",
    "compile:main": "cross-env BABEL_ENV=main babel ./src --out-dir ./lib/main --extensions \".ts\" --source-maps --config-file ../../babel.config.js",
    "compile:module": "cross-env BABEL_ENV=module babel ./src --out-dir ./lib/module --extensions \".ts\" --out-file-extension \".mjs\" --source-maps --config-file ../../babel.config.js",
    "compile:types": "tsc --build",
    "watch": "cross-env BABEL_ENV=browser rollup -c ../../rollup.config.js -w"
  },
  "dependencies": {
    "serialize-error": "8.1.0"
  },
  "peerDependencies": {
    "@babel/runtime": "< 8",
    "core-js": "< 4",
    "lodash": "< 5"
  }
}