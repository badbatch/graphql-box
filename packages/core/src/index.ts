import * as coreDefs from "./defs";
import Core from "./main";

export {
  Core as default,
  coreDefs,
};
