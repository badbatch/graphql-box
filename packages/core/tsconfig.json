{
  "extends": "../tsconfig.settings.json",
  "compilerOptions": {
    "outDir": "lib/types",
    "rootDir": "src"
  },
  "include": [
    "src"
  ]
}