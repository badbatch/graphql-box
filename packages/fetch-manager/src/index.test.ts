import { MaybeRawResponseData, RequestManagerDef } from "@graphql-box/core";
import {
  getRequestContext,
  getRequestData,
  parsedRequests,
  responses,
} from "@graphql-box/test-utils";
import fetchMock from "fetch-mock";
import { FetchManager } from ".";

const URL = "https://api.github.com/graphql";

describe("@graphql-box/fetch-manager >>", () => {
  let fetchManager: RequestManagerDef;

  describe("no batching >>", () => {
    let response: MaybeRawResponseData;

    beforeAll(async () => {
      fetchManager = await FetchManager.init({
        url: URL,
      });

      const body = { data: responses.singleTypeQuery.data };
      const headers = { "cache-control": "public, max-age=5" };
      fetchMock.post("*", { body, headers });

      response = await fetchManager.execute(
        getRequestData(parsedRequests.singleTypeQuery),
        {},
        getRequestContext(),
      );
    });

    afterAll(() => {
      fetchMock.restore();
    });

    it("correct request", () => {
      expect(fetchMock.lastCall()).toMatchSnapshot();
    });

    it("correct response data", () => {
      expect(response).toMatchSnapshot();
    });
  });

  describe("batching >>", () => {
    describe("single request >>", () => {
      let response: MaybeRawResponseData;

      beforeAll(async () => {
        jest.useFakeTimers();

        fetchManager = await FetchManager.init({
          batch: true,
          fetchTimeout: 10000,
          url: URL,
        });

        const requestData = getRequestData(parsedRequests.singleTypeQuery);

        const body = {
          batch: {
            [requestData.hash]: { data: responses.singleTypeQuery.data },
          },
        };

        const headers = { "cache-control": "public, max-age=5" };
        fetchMock.post("*", { body, headers });
        const promise = fetchManager.execute(requestData, {}, getRequestContext());
        jest.runOnlyPendingTimers();
        response = await promise;
      });

      afterAll(() => {
        fetchMock.restore();
      });

      it("correct request", () => {
        expect(fetchMock.lastCall()).toMatchSnapshot();
      });

      it("correct response data", () => {
        expect(response).toMatchSnapshot();
      });
    });

    describe("multiple requests >>", () => {
      let response: MaybeRawResponseData[];

      beforeAll(async () => {
        jest.useFakeTimers();

        fetchManager = await FetchManager.init({
          batch: true,
          fetchTimeout: 10000,
          url: URL,
        });

        const initialRequestData = getRequestData(parsedRequests.singleTypeQuerySet.initial);
        const updatedRequestData = getRequestData(parsedRequests.singleTypeQuerySet.updated);

        const body = {
          batch: {
            [initialRequestData.hash]: { data: responses.singleTypeQuerySet.initial.data },
            [updatedRequestData.hash]: { data: responses.singleTypeQuerySet.updated.data },
          },
        };

        const headers = { "cache-control": "public, max-age=5" };
        fetchMock.post("*", { body, headers });

        const promises = [
          fetchManager.execute(initialRequestData, {}, getRequestContext()),
          fetchManager.execute(updatedRequestData, {}, getRequestContext()),
        ];

        jest.runOnlyPendingTimers();
        response = await Promise.all(promises);
      });

      afterAll(() => {
        fetchMock.restore();
      });

      it("correct request", () => {
        expect(fetchMock.lastCall()).toMatchSnapshot();
      });

      it("correct response data", () => {
        expect(response[0]).toMatchSnapshot();
        expect(response[1]).toMatchSnapshot();
      });
    });
  });
});
