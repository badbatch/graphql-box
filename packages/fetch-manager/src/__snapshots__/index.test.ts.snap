// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@graphql-box/fetch-manager >> batching >> multiple requests >> correct request 1`] = `
Array [
  "https://api.github.com/graphql?requestId=f666cc27cdcfcecacb340b64fcf27506-70a245971c9c179bc5c30c3925e4881c",
  Object {
    "body": "{\\"batched\\":true,\\"context\\":{\\"boxID\\":\\"123456789\\",\\"operation\\":\\"query\\"},\\"request\\":{\\"f666cc27cdcfcecacb340b64fcf27506\\":\\"\\\\n    {\\\\n      organization(login: \\\\\\"facebook\\\\\\") {\\\\n        description\\\\n        login\\\\n        id\\\\n      }\\\\n    }\\\\n  \\",\\"70a245971c9c179bc5c30c3925e4881c\\":\\"\\\\n    {\\\\n      organization(login: \\\\\\"facebook\\\\\\") {\\\\n        email\\\\n        name\\\\n        url\\\\n        id\\\\n      }\\\\n    }\\\\n  \\"}}",
    "headers": Headers {
      "_headers": Object {
        "content-type": Array [
          "application/json",
        ],
      },
    },
    "method": "POST",
  },
]
`;

exports[`@graphql-box/fetch-manager >> batching >> multiple requests >> correct response data 1`] = `
Object {
  "data": Object {
    "organization": Object {
      "description": "We are working to build community through open source technology. NB: members must have two-factor auth.",
      "id": "MDEyOk9yZ2FuaXphdGlvbjY5NjMx",
      "login": "facebook",
    },
  },
  "headers": Headers {
    "_headers": Object {
      "cache-control": Array [
        "public, max-age=5",
      ],
      "content-length": Array [
        "414",
      ],
      "content-type": Array [
        "application/json",
      ],
    },
  },
}
`;

exports[`@graphql-box/fetch-manager >> batching >> multiple requests >> correct response data 2`] = `
Object {
  "data": Object {
    "organization": Object {
      "email": "",
      "id": "MDEyOk9yZ2FuaXphdGlvbjY5NjMx",
      "name": "Facebook",
      "url": "https://github.com/facebook",
    },
  },
  "headers": Headers {
    "_headers": Object {
      "cache-control": Array [
        "public, max-age=5",
      ],
      "content-length": Array [
        "414",
      ],
      "content-type": Array [
        "application/json",
      ],
    },
  },
}
`;

exports[`@graphql-box/fetch-manager >> batching >> single request >> correct request 1`] = `
Array [
  "https://api.github.com/graphql?requestId=ea1cbc58eca798225b53f59d95ffb23a",
  Object {
    "body": "{\\"batched\\":true,\\"context\\":{\\"boxID\\":\\"123456789\\",\\"operation\\":\\"query\\"},\\"request\\":{\\"ea1cbc58eca798225b53f59d95ffb23a\\":\\"\\\\n  {\\\\n    organization(login: \\\\\\"facebook\\\\\\") {\\\\n      description\\\\n      email\\\\n      login\\\\n      name\\\\n      url\\\\n      id\\\\n    }\\\\n  }\\\\n\\"}}",
    "headers": Headers {
      "_headers": Object {
        "content-type": Array [
          "application/json",
        ],
      },
    },
    "method": "POST",
  },
]
`;

exports[`@graphql-box/fetch-manager >> batching >> single request >> correct response data 1`] = `
Object {
  "data": Object {
    "organization": Object {
      "description": "We are working to build community through open source technology. NB: members must have two-factor auth.",
      "email": "",
      "id": "MDEyOk9yZ2FuaXphdGlvbjY5NjMx",
      "login": "facebook",
      "name": "Facebook",
      "url": "https://github.com/facebook",
    },
  },
  "headers": Headers {
    "_headers": Object {
      "cache-control": Array [
        "public, max-age=5",
      ],
      "content-length": Array [
        "315",
      ],
      "content-type": Array [
        "application/json",
      ],
    },
  },
}
`;

exports[`@graphql-box/fetch-manager >> no batching >> correct request 1`] = `
Array [
  "https://api.github.com/graphql?requestId=ea1cbc58eca798225b53f59d95ffb23a",
  Object {
    "body": "{\\"batched\\":false,\\"context\\":{\\"boxID\\":\\"123456789\\",\\"operation\\":\\"query\\"},\\"request\\":\\"\\\\n  {\\\\n    organization(login: \\\\\\"facebook\\\\\\") {\\\\n      description\\\\n      email\\\\n      login\\\\n      name\\\\n      url\\\\n      id\\\\n    }\\\\n  }\\\\n\\"}",
    "headers": Headers {
      "_headers": Object {
        "content-type": Array [
          "application/json",
        ],
      },
    },
    "method": "POST",
  },
]
`;

exports[`@graphql-box/fetch-manager >> no batching >> correct response data 1`] = `
Object {
  "data": Object {
    "organization": Object {
      "description": "We are working to build community through open source technology. NB: members must have two-factor auth.",
      "email": "",
      "id": "MDEyOk9yZ2FuaXphdGlvbjY5NjMx",
      "login": "facebook",
      "name": "Facebook",
      "url": "https://github.com/facebook",
    },
  },
  "headers": Headers {
    "_headers": Object {
      "cache-control": Array [
        "public, max-age=5",
      ],
      "content-length": Array [
        "268",
      ],
      "content-type": Array [
        "application/json",
      ],
    },
  },
}
`;
